packages/react-reconciler/src/ReactFiberWorkLoop.new.js

```diff
@@ -2993,64 +2999,74 @@ function flushRenderPhaseStrictModeWarningsInDEV() {
   }
 }
 
-function commitDoubleInvokeEffectsInDEV(
-  fiber: Fiber,
+function recursivelyTraverseAndDoubleInvokeEffectsInDEV(
+  root: FiberRoot,
+  parentFiber: Fiber,
+  isInStrictMode: boolean,
   hasPassiveEffects: boolean,
 ) {
-  if (__DEV__ && enableStrictEffects) {
-    // TODO (StrictEffects) Should we set a marker on the root if it contains strict effects
-    // so we don't traverse unnecessarily? similar to subtreeFlags but just at the root level.
-    // Maybe not a big deal since this is DEV only behavior.
+  let child = parentFiber.child;
+  while (child !== null) {
+    doubleInvokeEffectsInDEV(root, child, isInStrictMode, hasPassiveEffects);
+    child = child.sibling;
+  }
+}
 
+function doubleInvokeEffectsInDEV(
+  root: FiberRoot,
+  fiber: Fiber,
+  parentIsInStrictMode: boolean,
+  hasPassiveEffects: boolean,
+) {
+  const isStrictModeFiber = fiber.type === REACT_STRICT_MODE_TYPE;
+  const isInStrictMode = parentIsInStrictMode || isStrictModeFiber;
+  if (fiber.flags & PlacementDEV || fiber.tag === OffscreenComponent) {
     setCurrentDebugFiberInDEV(fiber);
-    invokeEffectsInDev(fiber, MountLayoutDev, invokeLayoutEffectUnmountInDEV);
-    if (hasPassiveEffects) {
-      invokeEffectsInDev(
-        fiber,
-        MountPassiveDev,
-        invokePassiveEffectUnmountInDEV,
-      );
+    if (isInStrictMode) {
+      disappearLayoutEffects(fiber);
```

Comment:
Factoring nit: All of these successive `if` branches include a `isInStrictMode` check. You can factor that out:

```js
if (isInStrictMode) {
  disappearLayoutEffects(...);
  disconnectPassiveEffect(...);
  reappearLayoutEfffects(...);
  disconnectPassiveEffect(...);
  reconnectPassiveEffects(...);
}
```

Also I don't think the setCurrentDebugFiberInDEV stuff is necessary anymore because disappearLayoutEffects et al already set it before they run the effects code.

Reply:
setCurrentDebugFiberInDEV is needed for DevTools. I'm not sure what the interaction between DevTools + React is in this case but DevTool tests start failing without it.